[WIP] Add Maskformer to the library

FrancescoSaverioZuppichini opened the pull request:
This WIP PR adds [MaskFormer](https://arxiv.org/abs/2107.06278) a new model for any segmentation task. This model ranks [on the top 10 of almost every task](https://paperswithcode.com/paper/per-pixel-classification-is-not-all-you-need)


A total of 8 pre-trained checkpoints will be available, which are the checkpoints discussed in the official MaskFormer paper. They are not yet available on the hub but tested locally. The weights are all pairs of combinations between the 4 Swin Transformer variants (tiny, small, base and large) and two datasets ([ade20k-150](https://groups.csail.mit.edu/vision/datasets/ADE20K/) and [coco-panoptic](https://cocodataset.org/#home))

TODOs
  - [x] converting script should be self-contained
    - currently, the job of downloading the weights and configuration files is outsourced to the end-user
  - [x] feature extractor
    - all the parameters needed are there, but the class is missing
    - the resizing needed is tricky, originally [ResizeShortestEdge](https://detectron2.readthedocs.io/en/latest/modules/data_transforms.html#detectron2.data.transforms.ResizeShortestEdge) from detectron is used 
    - all the post processing should be inside it
  - [x] backbone
    - recently we added [Swin Transformer](https://github.com/huggingface/transformers/pull/15085), the backbone should depend on that implementation
    - ported inside maskformer with the required changes
  - [x] loss
    - originally, the loss is computed by passing a list of dictionaries representing targets. However, this approach is subefficient since different targets may have different sizes making it hard to process the batch in a single go.
  - [x] padding
    - padding is handled in the forward pass using `NestedTensor`, this is a well know class used in a lot of implementations. This should be handled by the `FeatureExtractor` following [Niels implementation](https://github.com/huggingface/transformers/blob/c15bb3fe19b0b6c69a727812cdd3cd5597014667/src/transformers/models/detr/feature_extraction_detr.py#L633)
  - [x] auxiliary loss is not yet implemented
  - [x] output_hidden_states
  - [x] doc
  - [ ] tests:
    - [x] FeatureExtractor
    - [ ] MaskFormer
        
Currently, the model can be used as follows

```python
import torch

from transformers import (
    MaskFormerModel,
    MaskFormerForInstanceSegmentation,
    MaskFormerConfig,
    MaskFormerFeatureExtractor,
)

import numpy as np

feature_extractor = MaskFormerFeatureExtractor(do_resize=True)

inputs = feature_extractor(
    [np.zeros((3, 400, 1200)), np.zeros((3, 750, 384))],
    return_tensors="pt",
    pad_and_return_pixel_mask=True,
)

config = MaskFormerConfig()
mask_former = MaskFormerModel(config=config)
out = mask_former(**inputs)
# out contains the hidden states of each submodule
mask_former = MaskFormerForInstanceSegmentation(config=config)
out = mask_former(**inputs)
# out contains the logits
seg = feature_extractor.post_process_segmentation(out)
# get the instance panoptic mask + segments
seg = feature_extractor.post_process_panoptic_segmentation(out)

FrancescoSaverioZuppichini committed:
first forward pass

FrancescoSaverioZuppichini committed:
losses

FrancescoSaverioZuppichini committed:
naming, config

FrancescoSaverioZuppichini committed:
generic segmentation

FrancescoSaverioZuppichini committed:
little code refactor for matcher

FrancescoSaverioZuppichini committed:
little code refactor for matcher

FrancescoSaverioZuppichini committed:
little code refactor for matcher + some operations can be done inplace

FrancescoSaverioZuppichini committed:
types

FrancescoSaverioZuppichini committed:
types

FrancescoSaverioZuppichini committed:
types + refactor

FrancescoSaverioZuppichini committed:
panoptic segmentation

cannot finish it since we need the pretrain weights to detect something

FrancescoSaverioZuppichini committed:
comments

FrancescoSaverioZuppichini committed:
pretrained weigths

Currently not working, something after the decoder is wrong

FrancescoSaverioZuppichini committed:
doc in losses + refactor

FrancescoSaverioZuppichini committed:
weights are okay

FrancescoSaverioZuppichini committed:
weights transfer for semantic segmentation okay

however the output mask is weird

FrancescoSaverioZuppichini committed:
panoptic segmentation

main forward method is done, need to refactor how the output is returned

FrancescoSaverioZuppichini committed:
panoptic segmentation works

FrancescoSaverioZuppichini committed:
rewtring the original loss and fixing outputs and targets to have the same format

The original loss does a lot of non-fully vectorize operation and the original implementation is very hard to follow

FrancescoSaverioZuppichini committed:
refactor checkpoint converter

FrancescoSaverioZuppichini committed:
automati conversion of detectron config to ours

ready to insert all the pretrained models

FrancescoSaverioZuppichini committed:
automatically convert all the models from detectron to ours

FrancescoSaverioZuppichini committed:
dataset metadata from detectron config and test for pan seg

FrancescoSaverioZuppichini committed:
feature extractors parameters from detectron config

FrancescoSaverioZuppichini committed:
little doc

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15315). All of your documentation changes will be reflected on that endpoint.

FrancescoSaverioZuppichini committed:
imports

FrancescoSaverioZuppichini committed:
typos

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        \mathcal{L}_{\text{dice}(x, y) = 1 - \frac{2 * x * y }{x + y + 1}}
+
+    Args:
+        inputs (Tensor): A tensor representing a mask
```
```suggestion
        inputs (`torch.Tensor`): A tensor representing a mask
```
Putting it here once, but is applicable to every types in docstrings.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    inputs: Tensor, targets: Tensor, num_masks: int, alpha: float = 0.25, gamma: float = 2
+) -> Tensor:
+    r"""
+       Focal loss proposed in `Focal Loss for Dense Object Detection
+    <https://arxiv.org/abs/1708.02002>`_ originally used in RetinaNet. The loss is computed as follows
+
```
We use Markdown for the documentation, so that link needs to be converted.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    # TODO this .flatten seems to be unecessary because the shape is 2d
+    inputs: Tensor = inputs.flatten(1)
+    # TODO why 1 is not added to the number to avoid numerator = 0 in edge cases?
+    numerator: Tensor = 2 * torch.einsum("nc,mc->nm", inputs, targets)
```
This is just a transpose and matmul, so need to use einsum here.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    Returns:
+        Tensor: The computed loss
+    """
+    probs: Tensor = inputs.sigmoid()
+    ce_loss: Tensor = F.binary_cross_entropy_with_logits(inputs, targets, reduction="none")
+    p_t: Tensor = probs * targets + (1 - probs) * (1 - targets)
+    loss: Tensor = ce_loss * ((1 - p_t) ** gamma)
```
The `: Tensor` hurt readbility IMO. In a modeling file, pretty much everything is a tensor, so it's not an interesting type to mark.
`p_t` is too short a name, as is `ce_loss`. It would be more interesting to have longer names and be able to follow what those things represent IMO.

FrancescoSaverioZuppichini replied:
`p_t` comes from the paper equation, that is written in the doc of the function. I'd like to keep that name since it's consistent

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        inputs, torch.zeros_like(inputs), reduction="none"
+    )
+
+    focal_pos *= alpha
```
Since we are doing it on several lines, let's move the `((1 - prob) ** gamma)` here too and maybe the loss will fit in one line.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    )
+
+    focal_pos *= alpha
+    focal_neg *= 1 - alpha
```
Same comment here.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    focal_pos *= alpha
+    focal_neg *= 1 - alpha
+
+    loss: Tensor = torch.einsum("nc,mc->nm", focal_pos, targets) + torch.einsum("nc,mc->nm", focal_neg, (1 - targets))
```
Both are just transposes and matmuls.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+            cost_dice: This is the relative weight of the dice loss of the binary mask in the matching cost
+        """
+        super().__init__()
+        assert cost_class != 0 or cost_mask != 0 or cost_dice != 0, "all costs cant be 0"
```
No `assert` in a modeling file. We use tests and raise errors (as asserts can be skipped in some envs).

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+        Params:
+            outputs: This is a dict that contains at least these entries:
+                 "pred_logits": Tensor of dim [batch_size, num_queries, num_classes] with the classification logits
+                 "pred_masks": Tensor of dim [batch_size, num_queries, H_pred, W_pred] with the predicted masks
```
This should be a list and maybe put each key in bold instead of using quotes? Code should be in backticks.

sgugger replied:
This has been resolved but not treated.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    def __repr__(self):
+        head = "Matcher " + self.__class__.__name__
+        body = [
+            "cost_class: {}".format(self.cost_class),
+            "cost_mask: {}".format(self.cost_mask),
+            "cost_dice: {}".format(self.cost_dice),
```
No `format` in the code base, please use f-strings.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        # TODO this doc was copied by the authors
+        targets dicts must contain the key "labels" containing a tensor of dim [nb_target_masks]
+        """
+        assert "pred_logits" in outputs
```
No asserts please.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        """
+        assert "pred_logits" in outputs
+        pred_logits: Tensor = outputs["pred_logits"]
+        b, q, _ = pred_logits.shape
```
Better variable names please.

FrancescoSaverioZuppichini replied:
`b` and `q` were picked for consistency since they represent the same shape that is used later in `.rearrange`, do you think I should change the name there?

sgugger replied:
One-letter variable names are not readable, and as said before we should drop the requirement on einops anyway, so `rearrange` will go away.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        """Compute the losses related to the masks: the focal loss and the dice loss.
+        targets dicts must contain the key "masks" containing a tensor of dim [nb_target_masks, h, w]
+        """
+        assert "pred_masks" in outputs
```
No asserts please

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+    def get_loss(self, loss, outputs, targets, indices, num_masks):
+        loss_map = {"labels": self.loss_labels, "masks": self.loss_masks}
+        assert loss in loss_map, f"{loss} not in loss_map"
```
No asserts please.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+# TODO we could use our implementation of the swin transformer, that looks very similar to the authors' one,
+# with the downside of using more time to converting the weights
+# copied from original implementation
+class Mlp(nn.Module):
```
This class should be prefixed with the model name.

NielsRogge replied:
This class can be removed, along with all Swin-related classes below.

FrancescoSaverioZuppichini replied:
All the swin-related code will be removed

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    Returns:
+        windows: (num_windows*B, window_size, window_size, C)
+    """
+    B, H, W, C = x.shape
```
Better variable names please.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+def window_partition(x, window_size):
+    """
+    Args:
+        x: (B, H, W, C)
+        window_size (int): window size
+    Returns:
+        windows: (num_windows*B, window_size, window_size, C)
```
This is not very informative ;-)

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    Returns:
+        x: (B, H, W, C)
+    """
+    B = int(windows.shape[0] / (H * W / window_size / window_size))
+    x = windows.view(B, H // window_size, W // window_size, window_size, window_size, -1)
+    x = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(B, H, W, -1)
```
Same here.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+            x: input features with shape of (num_windows*B, N, C)
+            mask: (0/-inf) mask with shape of (num_windows, Wh*Ww, Wh*Ww) or None
+        """
+        B_, N, C = x.shape
+        qkv = self.qkv(x).reshape(B_, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
+        q, k, v = qkv[0], qkv[1], qkv[2]  # make torchscript happy (cannot use tensor as tuple)
```
Not very informative variable names.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+    def __init__(
+        self,
+        dim,
```
This should just take a config.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        drop=0.0,
+        attn_drop=0.0,
+        drop_path=0.0,
+        act_layer=nn.GELU,
```
We use string for activations and have a mapping for activation functions.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        self.window_size = window_size
+        self.shift_size = shift_size
+        self.mlp_ratio = mlp_ratio
+        assert 0 <= self.shift_size < self.window_size, "shift_size must in 0-window_size"
```
No assert please.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+            H, W: Spatial resolution of the input feature.
+            mask_matrix: Attention mask for cyclic shift.
+        """
+        B, L, C = x.shape
+        H, W = self.H, self.W
+        assert L == H * W, "input feature has wrong size"
```
Variable names + asserts

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+
+# copied from original implementation
+class PatchMerging(nn.Module):
```
Should be prefixed by the mode name.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+            x: Input feature, tensor size (B, H*W, C).
+            H, W: Spatial resolution of the input feature.
+        """
+        B, L, C = x.shape
+        assert L == H * W, "input feature has wrong size"
```
Variable names + asser

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+
+# copied from original implementation
+class BasicLayer(nn.Module):
```
Should be prefixed by the model name.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        self._freeze_stages()
+
+
+class BackboneMixin(nn.Module):
+    """This mixin defines a clear way to acces intermediate representation in the subclassing model.
+    A list of representations must be returned in the `forward` method, while their sizes in the `outputs_shape`.
+    """
+
+    def forward(self, *args, **kwargs) -> List[Tensor]:
+        raise NotImplemented
+
+    def outputs_shape(self) -> List[int]:
+        raise NotImplemented
```
If we decide to add some Mixins, this should be done in a specific PR just adding them IMO. It doesn't have anything specific to this one.

NielsRogge replied:
We could do this first maybe, to ensure we can leverage Swin Transformer in this PR.

NielsRogge replied:
This class can probably be defined in transformers/src/image_utils.py.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        return self.num_features
+
+
+class ConvLayer(nn.Sequential):
```
Building blocks should be prefixed with the model name.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        )
+
+
+class FPNLayer(nn.Module):
```
Building blocks should be prefixed with the model name.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        return down
+
+
+class FPNModel(nn.Module):
```
Building blocks should be prefixed with the model name.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    def forward(self, features: List[Tensor]) -> List[Tensor]:
+        fpn_features: List[Tensor] = []
+        last_feature: Tensor = features.pop()
+        x: Tensor = self.stem(last_feature)
```
`hidden` or `feature` is more informative than `x`.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+class MaskFormerPixelDecoder(nn.Module):
+    def __init__(self, *args, feature_size: int = 256, mask_feature_size: int = 256, **kwargs):
+        """Pixel Decoder Module proposed in `Per-Pixel Classification is Not All You Need for Semantic Segmentation <https://arxiv.org/abs/2107.06278>`_. It first run the backbone's feature into a Feature Pyramid Network creating a list of features map. Then, it projects the last one to the correct `mask_size`
```
Docstring should be written in Markdown.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    def __init__(self, *args, feature_size: int = 256, mask_feature_size: int = 256, **kwargs):
+        """Pixel Decoder Module proposed in `Per-Pixel Classification is Not All You Need for Semantic Segmentation <https://arxiv.org/abs/2107.06278>`_. It first run the backbone's feature into a Feature Pyramid Network creating a list of features map. Then, it projects the last one to the correct `mask_size`
+
+        Args:
+            feature_size (int, optional): The features (channels) of FPN feature maps. Defaults to 256.
+            mask_feature_size (int, optional): The features (channels) of the target masks size $C_{\epsilon}$ in the paper. Defaults to 256.
```
See any other model for the standard way we describe and type args with defaults.
Special behavior: we can't use `$ $` for inline math so the syntax is `\\( \\)`

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+
+# copied from original implementation, also practically equal to DetrSinePositionEmbedding
+class PositionEmbeddingSine(nn.Module):
```
Should be prefixed by the model name.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        return pos
+
+
+class MLP(nn.Sequential):
```
Isn't this the second `MLP`in this file?

NielsRogge replied:
This is the MLP used on top of the decoder outputs, I assume. In DETR, I called this class `DetrMLPPredictionHead`.
So here we could call it `MaskformerMLPPredictionHead`, maybe?

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    Returns:
+        Tensor: The upsampled tensor
+    """
+    _, _, h, w = like.shape
```
Better variable names please.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        )
+
+        losses = ["labels", "masks"]
+        # TODO I don't like too aggressive abbreviations in the naming
+        # what is 'ce'? It's cross entropy but it's not clear
+        # ask if we should follow a more exhaustive naming convention
```
Yes definitely. It's the `classification_loss` for instance if it's cross entropy, which is way clearer.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+
+@dataclass
+class MaskFormerForSemanticSegmentationOutput(MaskFormerOutput):
```
We usually don't use inheritance on model outputs types, to avoid having a sequence to follow. Just redefine the list of fields here.

NielsRogge replied:
Indeed, and please place it at the top of the modeling file.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        mask_classes: Tensor = outputs.pred_logits.softmax(dim=-1)[..., :-1]
+        # mask probs has shape [BATCH, QUERIES, HEIGHT, WIDTH]
+        mask_probs: Tensor = outputs.pred_masks.sigmoid()
+        # now we want to sum over the queries,
+        # $ out_{c,h,w} =  \sum_q p_{q,c} * m_{q,h,w} $
+        # where $ softmax(p) \in R^{q, c} $ is the mask classes
+        # and $ sigmoid(m) \in R^{q, h, w}$ is the mask probabilities
+        # b(atch)q(uery)c(lasses), b(atch)q(uery)h(eight)w(idth)
+        segmentation: Tensor = torch.einsum("bqc, bqhw -> bchw", mask_classes, mask_probs)
```
Very nice comment.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        return MaskFormerForSemanticSegmentationOutput(segmentation=segmentation, **outputs)
+
+
+class PanopticSegmentationSegment(TypedDict):
+    id: int
+    category_id: int
+    is_thing: bool
+    label: str
```
This looks out of scope of the modeling file. The model outputs logits, not final results.

FrancescoSaverioZuppichini replied:
`MaskFormerForPanopticSegmentation` is suppose to return a `panoptic` segmentation. Returning the right result will also remove responsibility from the extractor

sgugger replied:
No it's not supposed to return that.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    def remove_low_and_no_objects(
+        self, masks: Tensor, scores: Tensor, labels: Tensor
+    ) -> Tuple[Tensor, Tensor, Tensor]:
+        assert (
+            masks.shape[0] == scores.shape[0] == labels.shape[0]
+        ), "mask, scores and labels must have the same shape!"
```
No new assert.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        preds_logits: Tensor = outputs.pred_logits
+        preds_masks: Tensor = outputs.pred_masks
+
+        _, _, h, w = preds_masks.shape
```
Variable names are too short.

sgugger replied:
Again, this has been resolved but not addressed.

sgugger reviewed (commented):
Thanks a lot for your work on this. I've only left comments on the modeling and configuration file as I understand it's a work in progress. I tried to leave comments on the bits that didn't have TODO remove on them, or come directly from the original implementation.

In general:
- we don't use `assert` statements in modeling file, but tests that raise the appropriate error type.
- every docstring should be written in Markdown and follow our usual conventions. Docstrings are essential for the user-facing classes (model, model outputs) but not necessarily the building blocks where a comment in the code is more than enough. A docstring that says "forward pass" for the method `forward` is really not needed (either needs to be expanded but can most likely be removed)
- we don't usually using typing of each variable, and adding all the `Tensor` (everything is a tensor, surprise surprise), `bool` or `int` are not useful and only make the code harder to read IMO. You should focus on putting those for objects with a surprising type only. In most cases, picking good variable names will make the type obvious to the reader.
- one-letter variable names should be banned
- ever model class in the maskformer should be prefixed with `MaskFormer`. This allows researcher to easily mix and match blocks from different modeling files.

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+# coding=utf-8
+# Copyright 2021 Facebook AI Research and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 Facebook AI Research and The HuggingFace Inc. team. All rights reserved.
```

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+    r"""
+    This is the configuration class to store the configuration of a [`MaskFormer`]. It is used to instantiate a MaskFormer
+    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the "maskformer_swin_base_IN21k_384_bs16_160k_res640" architecture trained on ade20k-150
```
```suggestion
    defaults will yield a similar configuration to that of the "maskformer_swin_base_IN21k_384_bs16_160k_res640" architecture trained on ade20k-150
```
We'll have to decide on a default checkpoint. Let's make the model names not too long, we usually use "-" instead of "_" and don't include any capital letters.

Things like bs16 and 160k can probably be put in the model card as extra info.

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+
+        self.fpn_feature_size = fpn_feature_size
+        self.mask_feature_size = mask_feature_size
+        self.num_classes = num_classes
```
```suggestion
```
Each model in Transformers typically use `num_labels` for the number of classes. This is already defined in `PretrainedConfig`, so no need to define it here.

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+        dataset_metadata: DatasetMetadata = None,
+        fpn_feature_size: Optional[int] = 256,
+        mask_feature_size: Optional[int] = 256,
+        num_classes: Optional[int] = 150,
```
```suggestion
```
The number of labels can be set in the conversion script.

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+
+    def __init__(
+        self,
+        dataset_metadata: DatasetMetadata = None,
```
What's included in the metadata?

FrancescoSaverioZuppichini replied:
A list of class spec, defined as 

```python
@dataclass
class ClassSpec:
    is_thing: bool
    label: str
    color: Tuple[int, int, int]


@dataclass
class DatasetMetadata:
    classes: List[ClassSpec] = field(default_factory=list)
```

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+        num_classes: Optional[int] = 150,
+        no_object_weight: Optional[float] = 0.1,
+        num_queries: Optional[int] = 100,
+        swin_pretrain_img_size: Optional[int] = 384,
```
Models like ViT, DeiT and Swin use image_size regarding the image size they accept. Maybe for consistency, we can do the same here?
```suggestion
        image_size: Optional[int] = 384,
```

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+        no_object_weight: Optional[float] = 0.1,
+        num_queries: Optional[int] = 100,
+        swin_pretrain_img_size: Optional[int] = 384,
+        swin_in_channels: Optional[int] = 3,
```
```suggestion
        num_channels: Optional[int] = 3,
```
All existing vision models currently use num_channels regarding number of input channels.

FrancescoSaverioZuppichini replied:
I'll change it as soon as I bring our Swin Transformer in the implementation

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+        num_queries: Optional[int] = 100,
+        swin_pretrain_img_size: Optional[int] = 384,
+        swin_in_channels: Optional[int] = 3,
+        swin_patch_size: Optional[int] = 4,
+        swin_embed_dim: Optional[int] = 128,
+        swin_depths: Optional[List[int]] = None,
+        swin_num_heads: Optional[List[int]] = None,
+        swin_window_size: Optional[int] = 12,
+        swin_drop_path_rate: Optional[float] = 0.3,
```
So these are all attributes of the backbone. Wondering if we can just have a "backbone" attribute, similar to DETR.

However, that depends on whether we want to make it possible to use custom backbones, or Swin only. In the latter case, it makes sense to include all of these attributes.

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,150 @@
+        ce_weight: Optional[float] = 1.0,
+        mask_weight: Optional[float] = 20.0,
+        mask_classification: Optional[bool] = True,
+        max_position_embeddings: Optional[int] = 1024,
+        encoder_layers: Optional[int] = 6,
+        encoder_ffn_dim: Optional[int] = 2048,
+        encoder_attention_heads: Optional[int] = 8,
+        decoder_layers: Optional[int] = 6,
+        decoder_ffn_dim: Optional[int] = 2048,
+        decoder_attention_heads: Optional[int] = 8,
+        encoder_layerdrop: Optional[int] = 0.0,
+        decoder_layerdrop: Optional[int] = 0.0,
+        d_model: Optional[int] = 256,
+        dropout: Optional[int] = 0.1,
+        attention_dropout: Optional[int] = 0.0,
+        activation_dropout: Optional[int] = 0.0,
+        init_std: Optional[int] = 0.02,
+        init_xavier_std: Optional[int] = 1.0,
+        scale_embedding: Optional[int] = False,
+        auxiliary_loss: Optional[int] = False,
+        dilation: Optional[int] = False,
```
These parameters are not prefixed with "detr", whereas they are in the code below.

FrancescoSaverioZuppichini replied:
Thanks, nice catch :)

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+# coding=utf-8
+# Copyright 2021 Facebook AI Research The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 Facebook AI Research and The HuggingFace Inc. team. All rights reserved.
```

sgugger replied:
Same above. The comment is resolved but the suggestion was not applied...

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+from einops import rearrange
+from einops.einops import repeat
+from timm.models.layers import DropPath, to_2tuple, trunc_normal_
```
You can take a look at [Swin Transformer](https://github.com/huggingface/transformers/blob/e65bfc09718f132c76558d591be08c9751dd3ff2/src/transformers/models/swin/modeling_swin.py#L191) for an example.

We also typically add a # copied from statement to ensure consistency.

sgugger replied:
This has been resolved, but the imports are still here.

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+
+
+logger = logging.get_logger(__name__)
+import torch.distributed as dist
```
```suggestion
import torch.distributed as dist
```
Normally we also don't include any torch.distributed in the modeling file.

sgugger replied:
This has been resolved but no changes were made.

FrancescoSaverioZuppichini replied:
Do you think I can safely remove all the distributed related code?

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    return NestedTensor(tensor, mask)
+
+
+# copied from original implementation
+def is_dist_avail_and_initialized():
+    if not dist.is_available():
+        return False
+    if not dist.is_initialized():
+        return False
+    return True
```
```suggestion
```

FrancescoSaverioZuppichini committed:
Merge remote-tracking branch 'origin' into maskformer

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+class MaskFormerTransformerModule(nn.Module):
+    def __init__(self, in_features: int, config: MaskFormerConfig):
+        super().__init__()
+        self.position_embedder = PositionEmbeddingSine(num_pos_feats=config.hidden_size // 2, normalize=True)
+        self.queries_embedder = nn.Embedding(config.num_queries, config.hidden_size)
```
```suggestion
        self.position_embeddings = PositionEmbeddingSine(num_pos_feats=config.hidden_size // 2, normalize=True)
        self.queries_embeddings = nn.Embedding(config.num_queries, config.hidden_size)
```
Hmm we use the term `embeddings` for all other models currently, but to be honest embedder makes more sense :p wondering what @sgugger's opinion is here regarding naming.

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        self.position_embedder = PositionEmbeddingSine(num_pos_feats=config.hidden_size // 2, normalize=True)
+        self.queries_embedder = nn.Embedding(config.num_queries, config.hidden_size)
+        should_project = in_features != config.hidden_size
+        self.input_proj = (
```
```suggestion
        self.input_projection = (
```
Would use a clearer variable name.

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        super().__init__()
+        # + 1 because we add the "null" class
+        self.mask_classification = config.mask_classification
+        self.class_predictor = nn.Linear(config.hidden_size, config.num_classes + 1)
```
```suggestion
        self.class_predictor = nn.Linear(config.hidden_size, config.num_labels + 1)
```

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        # + 1 because we add the "null" class
+        self.mask_classification = config.mask_classification
+        self.class_predictor = nn.Linear(config.hidden_size, config.num_classes + 1)
+        self.mask_embedder = MLP(config.hidden_size, config.hidden_size, config.mask_feature_size)
```
```suggestion
        self.mask_embedder = MaskformerMLPPredictionHead(config.hidden_size, config.hidden_size, config.mask_feature_size)
```
If we are going with the name `MaskformerMLPPredictionHead`, we should also use it here.

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+            classes: Tensor = self.class_predictor(last_decoder_output)
+            out.update({"pred_logits": classes})
+        # sum up over the channels
+        binary_masks: Tensor = torch.einsum("bqc,   bchw -> bqhw", mask_embeddings, pixel_embbeddings)
```
```suggestion
        binary_masks: Tensor = torch.einsum("bqc, bchw -> bqhw", mask_embeddings, pixel_embbeddings)
```

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        self.class_predictor = nn.Linear(config.hidden_size, config.num_classes + 1)
+        self.mask_embedder = MLP(config.hidden_size, config.hidden_size, config.mask_feature_size)
+
+    def forward(self, decoder_outputs: Tuple[Tensor], pixel_embbeddings: Tensor) -> Dict[str, Tensor]:
```
```suggestion
    def forward(self, decoder_outputs: Tuple[Tensor], pixel_embeddings: Tensor) -> Dict[str, Tensor]:
```

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+        return out
+
+
+@dataclass
+class MaskFormerOutput(ModelOutput):
+    pred_logits: torch.FloatTensor
+    pred_masks: torch.FloatTensor = None
+    loss: Optional[torch.FloatTensor] = None
+    loss_dict: Optional[Dict] = None
```
I usually add the model output classes at the top of the modeling file.

NielsRogge commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    label: str
+
+
+@dataclass
+class MaskFormerForPanopticSegmentationOutput(MaskFormerForSemanticSegmentationOutput):
+    segments: List[PanopticSegmentationSegment] = None
```
Also here, please don't overwrite output classes, and place at the top of the modeling file.

NielsRogge reviewed (commented):
Did a first pass on the code, great work already!

Main things I noticed:

1) we can perhaps first open a PR regarding the addition of a general `BackboneMixIn`, such that we can use it in this PR.
2) the DETR decoder will have to be duplicated here, but we can add a `# copied from` statement.
3) agree with @sgugger regarding typing hints, we can add them but they sometimes make the code less readable (for instance, if there's already a docstring describing the arguments in detail).

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+from torch import Tensor, nn
+from torch.nn import functional as F
+
+from einops import rearrange
+from einops.einops import repeat
```
> Did a first pass on the code, great work already!
> 
> Main things I noticed:
> 
> 1. we can perhaps first open a PR regarding the addition of a general `BackboneMixIn`, such that we can use it in this PR.
> 2. the DETR decoder will have to be duplicated here, but we can add a `# copied from` statement.
> 3. agree with @sgugger regarding typing hints, we can add them but they sometimes make the code less readable (for instance, if there's already a docstring describing the arguments in detail).

Thanks for the review!

FrancescoSaverioZuppichini committed:
resolved conversations

FrancescoSaverioZuppichini committed:
resolved conflicts

FrancescoSaverioZuppichini committed:
config files (renamed mask_former -> maskformer)

FrancescoSaverioZuppichini committed:
feature extractor correctly returns pad images/targets

FrancescoSaverioZuppichini committed:
added feature_extractor file

FrancescoSaverioZuppichini committed:
removed unused postprocess functions from feature_extractor

FrancescoSaverioZuppichini committed:
feature extractor is working

FrancescoSaverioZuppichini committed:
feature extractor now returns 'labels' as a dict of 'pixel' and 'classes'

FrancescoSaverioZuppichini committed:
fixed types from the feature extractor

FrancescoSaverioZuppichini committed:
loss correctly computed using feature extractor labels, let's goooo

FrancescoSaverioZuppichini committed:
aux loss okayish

it looks like there is a little bug/idk in the implementation\nMoreover the self.weight dict only work for the basic one, not sure how they did it

FrancescoSaverioZuppichini committed:
Updated Readme.md

FrancescoSaverioZuppichini committed:
Updated maskformer.mdx

FrancescoSaverioZuppichini committed:
comments

FrancescoSaverioZuppichini committed:
fixed config

FrancescoSaverioZuppichini committed:
fixed conflicts

FrancescoSaverioZuppichini committed:
initial support for our swin model

the config is wrong, need to fix it

FrancescoSaverioZuppichini committed:
swin backbone works correctly with fixed pretrained size

FrancescoSaverioZuppichini committed:
swin weights conversion (patch + att layers)

FrancescoSaverioZuppichini committed:
swin weghts all converted but models outputs are different ...

FrancescoSaverioZuppichini committed:
swin buffers all converted but models outputs are different ...

FrancescoSaverioZuppichini committed:
fix last norm in backbone

FrancescoSaverioZuppichini committed:
minor refactor

weights still don't match'

FrancescoSaverioZuppichini committed:
problem found

commint before re-merge master in it

FrancescoSaverioZuppichini committed:
Merge branch 'master' of github.com:huggingface/transformers into maskformer

FrancescoSaverioZuppichini committed:
swin correctly converted

FrancescoSaverioZuppichini committed:
Feature Extractor Converted

FrancescoSaverioZuppichini committed:
Make style

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+    def forward(
+        self,
+        pixel_values: Tensor,
+    ):
```
I've run make style

FrancescoSaverioZuppichini committed:
swin padding done

FrancescoSaverioZuppichini committed:
swin backbone done

at each hidden state the correct spatial dimension is associated

FrancescoSaverioZuppichini committed:
feature extractor correctly stored

sgugger commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+        """
+
+        max_size = self._max_by_axis([list(image.shape) for image in pixel_values_list])
+        c, h, w = max_size
```
No one-letter variable names plese.

FrancescoSaverioZuppichini replied:
updated

sgugger commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+                if not is_torch_available():
+                    raise ImportError("Unable to convert output to PyTorch tensors format, PyTorch is not installed.")
+
+                encoded_inputs["labels"] = {
```
We shouldn't have a nested dict here for consistency with other models. The API should be:
```
features = FeatureExtracor(**inputs)
outputs = Model(**features)
```
e.g. the dict returned by the feature extractor should go in the model directly.

Models that expect several labels, have several arguments for each label (see DETR, or question answering models for instance). Here we should have `pixel_labels` and `class_labels`.

FrancescoSaverioZuppichini replied:
Okay. One issue is API clarity. E.g. if you pass `pixel_labels` you have to pass  `class_labels` as well, and vice-versa, you can enforce only with conditional raising. Passing a single dict solved it. In DETR `labels` is used. @NielsRogge

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,1638 @@
+)
+from ...modeling_utils import PreTrainedModel
+from ...utils import logging
+from ..detr import DetrConfig
+from ..detr.modeling_detr import DetrDecoder, DetrDecoderOutput
```
This still needs to be applied.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+    ModelOutput,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    is_scipy_available,
+    is_timm_available,
+    is_vision_available,
```
Not sure any of those should be used in the modeling file.

FrancescoSaverioZuppichini replied:
Removed

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+
+_CONFIG_FOR_DOC = "MaskFormerConfig"
+
+PREDICTIONS_MASKS_KEY = "preds_masks"
+PREDICTIONS_LOGITS_KEY = "preds_logits"
+TARGETS_MASKS_KEY = "pixel"
+TARGETS_LABELS_KEY = "classes"
```
What are those constants?

FrancescoSaverioZuppichini replied:
Easier to develop with. I'd like to fix the naming, are the currently used names okay to you?

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+import torch.distributed as dist
+
+
+_CONFIG_FOR_DOC = "MaskFormerConfig"
```
There should be a `_CHECKPOINT_FOR_DOC` and a `_FEATURE_EXTRACTOR_FOR_DOC` here. See other modeling files.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+
+
+@dataclass
+class MaskFormerOutput(ModelOutput):
```
This is missing a docstring to explain each field.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+
+@dataclass
+class MaskFormerOutput(ModelOutput):
+    preds_logits: torch.FloatTensor
```
Just `logits` for consistency with all other model outputs, since there is only one kind of logits.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+class MaskFormerOutput(ModelOutput):
+    preds_logits: torch.FloatTensor
+    preds_masks: torch.FloatTensor = None
+    loss: Optional[torch.FloatTensor] = None
```
The loss always needs to be the first field for backward compatibility reason and consistency with all other models.

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+
+
+@dataclass
+class MaskFormerForSemanticSegmentationOutput(ModelOutput):
```
Same comments as above plus what is `segmentation`?

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+
+
+@dataclass
+class MaskFormerForPanopticSegmentationOutput(ModelOutput):
```
Same comment as above plus what are `segmentation` and `segments`?

sgugger commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -0,0 +1,982 @@
+                      The expected keys in each dict depends on the losses applied, see each loss' doc
+        """
+        # TODO in theory here we can just take the `pred_masks` key
+        outputs_without_aux = {
+            PREDICTIONS_MASKS_KEY: outputs[PREDICTIONS_MASKS_KEY],
+            PREDICTIONS_LOGITS_KEY: outputs[PREDICTIONS_LOGITS_KEY],
+        }
```
Same as above.

sgugger commented on src/transformers/models/swin/modeling_swin.py:
```diff
@@ -167,12 +193,17 @@ def __init__(self, input_resolution, dim, norm_layer=nn.LayerNorm):
-    def forward(self, input_feature):
-        height, width = self.input_resolution
+    def forward(self, input_feature, height, width):
+        # height, width = self.input_resolution
```
Needs to be cleaned up.

FrancescoSaverioZuppichini replied:
now all functions that needs to know the spatial dimension, a tuple `input_dimensions` is always passed

sgugger commented on src/transformers/models/swin/modeling_swin.py:
```diff
@@ -42,6 +44,16 @@
 ]
 
 
+@dataclass
+class SwinModelOutputWithPooling(BaseModelOutputWithPooling):
+    hidden_states_original_spatial_dimensions: Tuple[Tuple[int, int]] = None
+
+
+@dataclass
+class SwinBaseModelOutput(BaseModelOutput):
+    hidden_states_original_spatial_dimensions: Tuple[Tuple[int, int]] = None
```
I don't understand why those new outputs are necessary and what this new field represent. As said before, please redefine the full class (and not rely on inheritance) and include a full docstring.

FrancescoSaverioZuppichini replied:
Okay I thought we could inherit from the internal ones (like we did with `ModelOutput`). I need to know the spatial dimensions of each feature map to resize them later on (in the Backbone)

sgugger commented on src/transformers/models/swin/modeling_swin.py:
```diff
@@ -416,17 +448,23 @@ def __init__(self, config, dim, input_resolution, num_heads, shift_size=0):
         hidden_states = hidden_states.view(batch_size, height, width, channels)
-
+        # pad feature maps to multiples of window size
+        pad_l = pad_t = 0
+        pad_r = (self.window_size - width % self.window_size) % self.window_size
+        pad_b = (self.window_size - height % self.window_size) % self.window_size
+        pad_value = (0, 0, pad_l, pad_r, pad_t, pad_b)
```
if l, r, b and t represent left, rgiht, bottom and top, they should be changed to that. The code would be more readable with those.

FrancescoSaverioZuppichini replied:
fixed, all pad related code was also moved to methods called `maybe_pad`

sgugger reviewed (commented):
I don't fully understand why some conversations were resolved without applying the suggestion in them or applying the suggestion.

The changes to the Swin model would need to go in a separate PR, with explanations on why they are necessary (for instance I don't get why we need to output those dimensions and can't infer them from the hidden states).

The feature extractor should be added to the main init.

As discussed:
- we shouldn't introduce a dependency on `einops` just for this model
- variable names should be descriptive and not one-letter long
- the documentation should follow the standardized format
- there is no need for all the `Tensor` type annotations
- the model outputs should be limited to tensors pre-last activations and any kind of post-processing should be either done in an example of use or the feature extractor decode method.

FrancescoSaverioZuppichini commented on src/transformers/models/swin/modeling_swin.py:
```diff
@@ -478,6 +519,7 @@ def __init__(self, config, dim, input_resolution, depth, num_heads, drop_path, d
         super().__init__()
         self.config = config
         self.dim = dim
+        self.input_resolution = input_resolution
```
Good catch, nowhere, it was a leftover from our original implentation

NielsRogge commented on docs/source/model_doc/maskformer.mdx:
```diff
@@ -0,0 +1,59 @@
+
+## MaskFormer specific outputs
+
+### Base
+
+[[autodoc]] models.maskformer.modeling_maskformer.MaskFormerOutput
+
+### Segmentation
+
+[[autodoc]] models.maskformer.modeling_maskformer.MaskFormerForSemanticSegmentationOutput
+
+### Panoptic Segmentation
+
+[[autodoc]] models.maskformer.modeling_maskformer.MaskFormerForPanopticSegmentationOutput
```
Indeed, you can remove the ones with 3 #

FrancescoSaverioZuppichini replied:
Done

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+        swin_in_channels: Optional[int] = 3,
+        swin_patch_size: Optional[int] = 4,
+        swin_embed_dim: Optional[int] = 128,
+        swin_depths: Optional[List[int]] = None,
+        swin_num_heads: Optional[List[int]] = None,
```
Done

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -1098,6 +1100,17 @@
+            "MaskFormerLayer",
+            "MaskFormerModel",
+            "MaskFormerPreTrainedModel",
+            "load_tf_weights_in_maskformer",
```
Indeed, the `load_tf_weights_in_maskformer` function is only used when porting a TF model to PyTorch.

FrancescoSaverioZuppichini replied:
Removed, it was created by the cl

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+class ClassSpec:
+    is_thing: bool
+    label: str
+    color: Tuple[int, int, int]
```
Yes, I don't know where to store this information and if it's needed here or it's best to have it somewhere else (e.g. the pipeline?)

NielsRogge replied:
Yes, although I've seen some segmentation datasets define a so-called "palette", that maps each class to a specific color. Example can be found here: https://github.com/open-mmlab/mmsegmentation/blob/4b905cbe2fb2a8afbb7e91c5453aa5fd8ac6b060/mmseg/core/evaluation/class_names.py#L114

I was also wondering where we should define these palettes in the library.

NielsRogge commented on src/transformers/models/maskformer/__init__.py:
```diff
@@ -0,0 +1,59 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

FrancescoSaverioZuppichini replied:
Thanks, fixed

NielsRogge commented on src/transformers/models/maskformer/__init__.py:
```diff
@@ -0,0 +1,59 @@
+
+
+if is_torch_available():
+    _import_structure["modeling_maskformer"] = [
+        "MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP",
+        "MaskFormerForPanopticSegmentation",
+        "MaskFormerForPanopticSegmentationOutput",
+        "MaskFormerForSemanticSegmentation",
+        "MaskFormerForSemanticSegmentationOutput",
+        "MaskFormerModel",
+    ]
```
```suggestion
    _import_structure["modeling_maskformer"] = [
        "MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP",
        "MaskFormerForPanopticSegmentation",
        "MaskFormerForSemanticSegmentation",
        "MaskFormerModel",
    ]
```
Output classes should not be here.

FrancescoSaverioZuppichini replied:
Removed

NielsRogge commented on src/transformers/models/maskformer/__init__.py:
```diff
@@ -0,0 +1,59 @@
+        # TODO code it!
+        # from .feature_extraction_maskformer import MaskFormerFeatureExtractor
+        pass
+    if is_torch_available():
+        from .modeling_maskformer import (
+            MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP,
+            MaskFormerForPanopticSegmentation,
+            MaskFormerForPanopticSegmentationOutput,
+            MaskFormerForSemanticSegmentation,
+            MaskFormerForSemanticSegmentationOutput,
+            MaskFormerModel,
+        )
```
```suggestion
    if is_torch_available():
        from .modeling_maskformer import (
            MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP,
            MaskFormerForPanopticSegmentation,
            MaskFormerForSemanticSegmentation,
            MaskFormerModel,
        )
```
Output classes should not be here.

FrancescoSaverioZuppichini replied:
Removed

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+
+
+MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    # "facebook/detr-resnet-50": "https://huggingface.co/facebook/detr-resnet-50/resolve/main/config.json",
+    # See all MaskFormer models at https://huggingface.co/models?filter=detr
```
Yes, something like `facebook/maskformer-swin-base-ade-640`.

FrancescoSaverioZuppichini replied:
sure, is there a preferred format? E.g. `<MODEL_MAME>-<BACKBONE>-<DATASET>-<INPUT_SIZE>`

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+    r"""
+    This is the configuration class to store the configuration of a [`MaskFormer`]. It is used to instantiate a
+    MaskFormer model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the "askformer-swin-base-640"
```
```suggestion
    configuration with the defaults will yield a similar configuration to that of the "maskformer-swin-base-ade-640"
```
Maybe we can call it like that.

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+    r"""
+    This is the configuration class to store the configuration of a [`MaskFormer`]. It is used to instantiate a
+    MaskFormer model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the "askformer-swin-base-640"
+    architecture trained on ade20k-150
```
```suggestion
    configuration with the defaults will yield a similar configuration to that of the "maskformer-swin-base-ade-640"
    architecture.
```

FrancescoSaverioZuppichini replied:
Updated the model name

NielsRogge commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+
+    def __init__(
+        self,
+        dataset_metadata: DatasetMetadata = None,
```
The `id2label` and `label2id` mappings are typically set when converting checkpoints, in the conversion script.

FrancescoSaverioZuppichini replied:
`id2label` and `label2id` are correctly set in the config file

NielsRogge commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+from __future__ import annotations
+
+import enum
+import logging
+from argparse import ArgumentParser
+from ast import Tuple
+from dataclasses import dataclass
+from pathlib import Path
+from pprint import pformat, pprint
+from typing import Any, Dict, Iterator, List, Set
+
+import pytorch_lightning as pl
+import torch
+import torchvision.transforms as T
+from PIL import Image
+from torch import Tensor
```
Yeah that's a huge number of imports. Please remove any unnecessary ones.

Also, eventually you can remove `torchvision.transforms as T` and use the feature extractor instead.

FrancescoSaverioZuppichini replied:
PyTorch Lightning was used to seed everything, it is not used anymore -> removed
 `ast` was wrongly imported by vs-code auto import extension -> removed

NielsRogge commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+    MaskFormerForSemanticSegmentationOutput,
+    MaskFormerModel,
+)
+from transformers.models.maskformer.configuration_maskformer import ClassSpec, DatasetMetadata
```
```suggestion
from transformers.models.maskformer.configuration_maskformer import ClassSpec, DatasetMetadata
```
To be removed if we add the additional mappings to `PretrainedConfig`.

NielsRogge commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+logger.addHandler(logging.StreamHandler())
+from dataclasses import dataclass, field
+
+from torch import Tensor
```
This can be removed imo, having `Tensor` type annotations are not that useful.

NielsRogge commented on docs/source/model_doc/maskformer.mdx:
```diff
@@ -0,0 +1,59 @@
+
+### Base
+
+[[autodoc]] models.maskformer.modeling_maskformer.MaskFormerOutput
```
I would call the model Maskformer everywhere, so:

* MaskformerOutput
* MaskformerForSemanticSegmentation
* MaskformerForPanopticSegmentation

etc.

FrancescoSaverioZuppichini replied:
The name on the paper is`MaskFormer`, may I ask you why you wish to change it to `Maskformer`?

NielsRogge replied:
Yeah so we have `ViTModel` and `ViTForImageClassification` for instance. I actually rather wanted to have `VitModel` and `VitForImageClassification` (it's not great with small and capital letters for people to type).

There's also:
* SegFormer, but the model is called `SegformerModel` in the library. 
* ConvNeXT, which will be called`ConvnextModel`.
* DETR, which is called `DetrModel`.

Just a matter of easy typing for me :) cc @sgugger

sgugger replied:
The same way we have `BertModel`, `RobertaModel` but the models are called BERT and RoBERTa. In general, we prefer simple casing for the class prefix (like `Maskformer` here). When we don't do this (like for `ViT`) it results in lot of small issues with typing.

FrancescoSaverioZuppichini replied:
Thank you for the clarification, I'll fix it

FrancescoSaverioZuppichini commented on README.md:
```diff
@@ -279,6 +279,7 @@ Min, Patrick Lewis, Ledell Wu, Sergey Edunov, Danqi Chen, and Wen-tau Yih.
 1. **[LXMERT](https://huggingface.co/docs/transformers/model_doc/lxmert)** (from UNC Chapel Hill) released with the paper [LXMERT: Learning Cross-Modality Encoder Representations from Transformers for Open-Domain Question Answering](https://arxiv.org/abs/1908.07490) by Hao Tan and Mohit Bansal.
 1. **[M2M100](https://huggingface.co/docs/transformers/model_doc/m2m_100)** (from Facebook) released with the paper [Beyond English-Centric Multilingual Machine Translation](https://arxiv.org/abs/2010.11125) by Angela Fan, Shruti Bhosale, Holger Schwenk, Zhiyi Ma, Ahmed El-Kishky, Siddharth Goyal, Mandeep Baines, Onur Celebi, Guillaume Wenzek, Vishrav Chaudhary, Naman Goyal, Tom Birch, Vitaliy Li ...
 1. **[MarianMT](https://huggingface.co/docs/transformers/model_doc/marian)** Machine translation models trained using [OPUS](http://opus.nlpl.eu/) data by Jörg Tiedemann. The [Marian Framework](https://marian-nmt.github.io/) is being developed by the Microsoft Translator Team.
+1. **[MaskFormer](https://huggingface.co/docs/transformers/model_doc/maskformer)** (from Meta and UIUC) released with the paper [Per-Pixel Classification is Not All You Need for Semantic Segmentation](https://arxiv.org/abs/2107.06278) by Bowen Cheng, Alexander G. Schwing, Alexander Kirillov
```
Thanks, why all the other models are not in the `master` doc yet?

sgugger replied:
The other models are in the stable doc because there has been a release. Once your model is merged, it won't be in the stable doc, only the master doc and the link as is will 404. The links are simplified when we make a release.

FrancescoSaverioZuppichini replied:
Thank you very much for the clarification

FrancescoSaverioZuppichini commented on README.md:
```diff
@@ -279,6 +279,7 @@ Min, Patrick Lewis, Ledell Wu, Sergey Edunov, Danqi Chen, and Wen-tau Yih.
+1. **[MaskFormer](https://huggingface.co/docs/transformers/model_doc/maskformer)** (from Meta and UIUC) released with the paper [Per-Pixel Classification is Not All You Need for Semantic Segmentation](https://arxiv.org/abs/2107.06278) by Bowen Cheng, Alexander G. Schwing, Alexander Kirillov
 1. **[MBart](https://huggingface.co/docs/transformers/model_doc/mbart)** (from Facebook) released with the paper [Multilingual Denoising Pre-training for Neural Machine Translation](https://arxiv.org/abs/2001.08210) by Yinhan Liu, Jiatao Gu, Naman Goyal, Xian Li, Sergey Edunov, Marjan Ghazvininejad, Mike Lewis, Luke Zettlemoyer.
 1. **[MBart-50](https://huggingface.co/docs/transformers/model_doc/mbart)** (from Facebook) released with the paper [Multilingual Translation with Extensible Multilingual Pretraining and Finetuning](https://arxiv.org/abs/2008.00401) by Yuqing Tang, Chau Tran, Xian Li, Peng-Jen Chen, Naman Goyal, Vishrav Chaudhary, Jiatao Gu, Angela Fan.
 1. **[Megatron-BERT](https://huggingface.co/docs/transformers/model_doc/megatron-bert)** (from NVIDIA) released with the paper [Megatron-LM: Training Multi-Billion Parameter Language Models Using Model Parallelism](https://arxiv.org/abs/1909.08053) by Mohammad Shoeybi, Mostofa Patwary, Raul Puri, Patrick LeGresley, Jared Casper and Bryan Catanzaro.
```
I've run `make fix-copies` from the top root

FrancescoSaverioZuppichini commented on docs/source/model_doc/maskformer.mdx:
```diff
@@ -0,0 +1,59 @@
+
+## MaskFormerModel
+
+[[autodoc]] MaskFormerModel - forward
```
done

FrancescoSaverioZuppichini commented on docs/source/model_doc/maskformer.mdx:
```diff
@@ -0,0 +1,59 @@
+
+## MaskFormerForSemanticSegmentation
+
+[[autodoc]] MaskFormerForSemanticSegmentation - forward
```
done

FrancescoSaverioZuppichini commented on docs/source/model_doc/maskformer.mdx:
```diff
@@ -0,0 +1,59 @@
+
+## MaskFormerForPanopticSegmentation
+
+[[autodoc]] MaskFormerForPanopticSegmentation - forward
```
done

FrancescoSaverioZuppichini commented on src/transformers/__init__.py:
```diff
@@ -244,6 +244,7 @@
     "models.lxmert": ["LXMERT_PRETRAINED_CONFIG_ARCHIVE_MAP", "LxmertConfig", "LxmertTokenizer"],
     "models.m2m_100": ["M2M_100_PRETRAINED_CONFIG_ARCHIVE_MAP", "M2M100Config"],
     "models.marian": ["MarianConfig"],
+    "models.maskformer": ["MASK_FORMER_PRETRAINED_CONFIG_ARCHIVE_MAP", "MaskFormerConfig", "MaskFormerTokenizer"],
```
Thanks

FrancescoSaverioZuppichini commented on src/transformers/__init__.py:
```diff
@@ -1098,6 +1100,17 @@
+            "MASK_FORMER_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "MaskFormerForPanopticSegmentation",
+            "MaskFormerForSemanticSegmentation",
+            "MaskFormerLayer",
```
Removed

FrancescoSaverioZuppichini commented on src/transformers/__init__.py:
```diff
@@ -2433,6 +2446,7 @@
     from .models.lxmert import LXMERT_PRETRAINED_CONFIG_ARCHIVE_MAP, LxmertConfig, LxmertTokenizer
     from .models.m2m_100 import M2M_100_PRETRAINED_CONFIG_ARCHIVE_MAP, M2M100Config
     from .models.marian import MarianConfig
+    from .models.maskformer import MASK_FORMER_PRETRAINED_CONFIG_ARCHIVE_MAP, MaskFormerConfig, MaskFormerTokenizer
```
Removed

FrancescoSaverioZuppichini commented on src/transformers/__init__.py:
```diff
@@ -2610,6 +2624,7 @@
         from .models.led import LEDTokenizerFast
         from .models.longformer import LongformerTokenizerFast
         from .models.lxmert import LxmertTokenizerFast
+        from .models.maskformer import MaskFormerTokenizerFast
```
Removed

FrancescoSaverioZuppichini commented on src/transformers/__init__.py:
```diff
@@ -3136,6 +3152,15 @@
+        from .models.maskformer import (
+            MASK_FORMER_PRETRAINED_MODEL_ARCHIVE_LIST,
+            MaskFormerForPanopticSegmentation,
+            MaskFormerForSemanticSegmentation,
+            MaskFormerLayer,
+            MaskFormerModel,
+            MaskFormerPreTrainedModel,
+            load_tf_weights_in_maskformer,
```
Removed

FrancescoSaverioZuppichini commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -280,6 +281,7 @@
     [
         # Model for Image Segmentation mapping
         ("detr", "DetrForSegmentation"),
+        ("maskformer", ("MaskFormerForSemanticSegmentation", "MaskFormerForPanopticSegmentation")),
```
Thanks for the comment. I am not following, it depends on how segmentation is defined, if by segmentation we mean return an image with class labels on it then can just leave `MaskFormerForSemanticSegmentation` inside the auto mapping. For example, `DetrForSegmentation` doesn't return a segmentation map that matches the previous definition, thus the user can't use it alone.

FrancescoSaverioZuppichini replied:
Update to only `MaskFormerForInstanceSegmentation`

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    from .configuration_maskformer import MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP, MaskFormerConfig
+
+    if is_vision_available():
+        # TODO code it!
+        # from .feature_extraction_maskformer import MaskFormerFeatureExtractor
+        pass
```
Indeed :)

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/configuration_maskformer.py:
```diff
@@ -0,0 +1,149 @@
+        self.swin_in_channels = swin_in_channels
+        self.swin_patch_size = swin_patch_size
+        self.swin_embed_dim = swin_embed_dim
+        self.swin_depths = swin_depths or [2, 2, 18, 2]
+        self.swin_num_heads = swin_num_heads or [4, 8, 16, 32]
```
Moved to default parameters

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+
+
+StateDict = Dict[str, Tensor]
+# easier to use pythong logging instead of going trough the hf utility logging file
+# main issue there is the polluted namespace so I can't call `logging` directly
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler())
```
I've tried, using 

```
from transformers.utils import logging


logging.set_verbosity_info()
logger = logging.get_logger(__name__)
```

results in no output, could you please share the correct configuration?

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+from torch import Tensor
+
+
+class TrackedStateDict:
```
It's explained in the docstring. Is it unclear? I need to track with keys have been accessed so I know for sure everything was updated

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+            dst_state_dict[f"{dst_prefix}.hidden_states_norms.3.weight"],
+            dst_state_dict[f"{dst_prefix}.model.layernorm.weight"],
+        )
+        # dst_state_dict[f"{dst_prefix}.hidden_states_norms.3.weight"].copy_(
+        #     dst_state_dict[f"{dst_prefix}.model.layernorm.weight"]
+        # )
+
+        # dst_state_dict[f"{dst_prefix}.hidden_states_norms.3.bias"].copy_(
+        #     dst_state_dict[f"{dst_prefix}.model.layernorm.bias"]
+        # )
```
Definitely

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+            ), config, checkpoint
+
+
+checkpoints_dir = Path("/home/zuppif/Documents/Work/hugging_face/maskformer/weights")
+config_dir = Path("/home/zuppif/Documents/Work/hugging_face/maskformer/MaskFormer/configs")
```
Yes, there way used for local testing. Removed

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Feature extractor class for DETR."""
```
Fixed

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/convert_maskformer_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,768 @@
+        converted: MaskFormerModel = converter()
+        test(converter.original_model, converted)
+
+    # pl.seed_everything(42)
```
Done

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+import PIL
```
Removed

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+        size = get_size(image.size, size, max_size)
+        rescaled_image = self.resize(image, size=size)
+        # pil image have inverted w, h
+        w, h = size
```
(Hopefully) update every occurrence of `h,w` with their corresponding full names. Keep in mind that the code was mostly copied from DETR

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+        """
+        Normalize the image with a certain mean and std.
+
+        If given, also normalize the target bounding boxes based on the size of the image.
```
Thanks. I'll update all the docstrings once the codebase is fixed

FrancescoSaverioZuppichini commented on src/transformers/models/maskformer/feature_extraction_maskformer.py:
```diff
@@ -0,0 +1,373 @@
+        self,
+        images: ImageInput,
+        annotations: Union[List[Dict], List[List[Dict]]] = None,
+        pad_and_return_pixel_mask: Optional[bool] = True,
```
This was copied from DETR, if we change it here we should change there as well otherwise we won't have consistency across FeatureExtractors. To my understanding a FeatureExtract is not a Tokenizer @NielsRogge

FrancescoSaverioZuppichini committed:
minor changes

FrancescoSaverioZuppichini committed:
imports cleanup

FrancescoSaverioZuppichini committed:
addressed Niels comments

FrancescoSaverioZuppichini committed:
doc string for new swin outputs'

FrancescoSaverioZuppichini committed:
docstring for maskformer

FrancescoSaverioZuppichini committed:
replace_return_docstrings creates an error

FrancescoSaverioZuppichini committed:
detr config causes error

FrancescoSaverioZuppichini committed:
nested config works

FrancescoSaverioZuppichini committed:
removed num_queries from config

user forced to use the transformer_decoder_config

FrancescoSaverioZuppichini committed:
minor changes

FrancescoSaverioZuppichini committed:
typo

FrancescoSaverioZuppichini committed:
MaskFormerPretrainedModel & loss

moved the loss computation away for MaskFormerModel

FrancescoSaverioZuppichini committed:
moved runtime parameters to forward

FrancescoSaverioZuppichini committed:
fixed nested configs in MaskFormerConfig

FrancescoSaverioZuppichini committed:
moved all the loss related code to MaskFormerPretrainedModel

FrancescoSaverioZuppichini committed:
loss in MaskFormerForPretraining

FrancescoSaverioZuppichini committed:
changed name in the outputs to match the discussion

FrancescoSaverioZuppichini committed:
removed MaskFormerForSemanticSegmentation

post processing moved into feature extractor

FrancescoSaverioZuppichini committed:
removed MaskFormerForPanopticSegmentation

post processing moved into feature extractor

FrancescoSaverioZuppichini committed:
removed global outputs variables

FrancescoSaverioZuppichini committed:
minor refactor

FrancescoSaverioZuppichini committed:
add forward in MaskFormerForPretraining

FrancescoSaverioZuppichini committed:
refactor + docstring

FrancescoSaverioZuppichini committed:
mask former init weights method

FrancescoSaverioZuppichini committed:
the submodules have a custom output

Added output_hidden_states

FrancescoSaverioZuppichini committed:
add outputs for all the submodule + MaskFormerModel is return all the hidden_states

FrancescoSaverioZuppichini committed:
Huge refactor

Now each submodule has its output, MaskFormerForInstanceSegmentation computes the loss and the logits

FrancescoSaverioZuppichini committed:
Documentation for Outputs

FrancescoSaverioZuppichini committed:
Documentation for all Models

FrancescoSaverioZuppichini committed:
Updated root.__init__

FrancescoSaverioZuppichini committed:
removed datasetmetadata

the features extractor now receives a is_thing_map like in detr

FrancescoSaverioZuppichini committed:
typos

FrancescoSaverioZuppichini committed:
basic tests

Commit:
Merge branch 'master' into maskformer

FrancescoSaverioZuppichini committed:
brutaly removed some of the doc in swin to make the model works

FrancescoSaverioZuppichini committed:
dropped einops

FrancescoSaverioZuppichini committed:
dropped TypedDict

minor changes in the FeatureExtractor's doc

FrancescoSaverioZuppichini committed:
fixed doc

The old  were still there

FrancescoSaverioZuppichini committed:
fixed imports and dummy

FrancescoSaverioZuppichini committed:
removed F.

FrancescoSaverioZuppichini committed:
removed unused imports

FrancescoSaverioZuppichini committed:
Merge remote-tracking branch 'origin' into maskformer

FrancescoSaverioZuppichini committed:
maskformer test skeletron

FrancescoSaverioZuppichini committed:
maskformer added to _toctree.yml

FrancescoSaverioZuppichini committed:
refactor the conversion script to match ours changes

FrancescoSaverioZuppichini committed:
minor refactor

FrancescoSaverioZuppichini committed:
add semantic segmentation postprocess

FrancescoSaverioZuppichini committed:
logger

FrancescoSaverioZuppichini committed:
minor changes

FrancescoSaverioZuppichini committed:
minor changes

FrancescoSaverioZuppichini committed:
test_maskformer_model passes

FrancescoSaverioZuppichini committed:
run make style

FrancescoSaverioZuppichini committed:
Correcetly reshaped the inputs for DETR

DetrDecoder expects inputs to be batch x time x channel, while we were passing time x batch x channel. Interesting enough, the model was working properly when batch_size = 1.

FrancescoSaverioZuppichini committed:
feature extractor tests + make style

FrancescoSaverioZuppichini committed:
basic tests for maskformer

FrancescoSaverioZuppichini committed:
make style

FrancescoSaverioZuppichini committed:
fixed import in maskformer

FrancescoSaverioZuppichini committed:
added return_dict

FrancescoSaverioZuppichini committed:
conversion script working

FrancescoSaverioZuppichini committed:
added detr inside maskformer

FrancescoSaverioZuppichini committed:
added swin inside maskformer

FrancescoSaverioZuppichini committed:
removed #Copied from

FrancescoSaverioZuppichini committed:
swin now matches master

FrancescoSaverioZuppichini committed:
update model auto

FrancescoSaverioZuppichini committed:
update model auto

FrancescoSaverioZuppichini committed:
update swin in maskformer

FrancescoSaverioZuppichini committed:
update swin in maskformer

FrancescoSaverioZuppichini committed:
make fixies

FrancescoSaverioZuppichini committed:
fixed the docstring destroyed by make-fixies

FrancescoSaverioZuppichini committed:
typos

FrancescoSaverioZuppichini committed:
typos

FrancescoSaverioZuppichini committed:
typos
